src/frontend/src/binder/expr/function.rs

```diff
@@ -672,6 +672,20 @@ impl Binder {
             )
         }
 
+        fn reduce(r#type: ExprType) -> Handle {
```

Comment:
There is variadic function support: #12178

> The expressions must all be convertible to a common data type, which will be the type of the result (see [Section 10.5](https://www.postgresql.org/docs/current/typeconv-union-case.html) for details).

In the frontend this can be achieved with [`align_types`](https://github.com/risingwavelabs/risingwave/blob/3579e74ea9194170fcec8f1f74c4f4aac6b4b2fd/src/frontend/src/expr/type_inference/func.rs#L319).

Reply:
I just checked the function currently using the variadic input. The input type of the function implementations are all `f(row: impl Row)` like `array`
```
#[function("array(...) -> list")]
fn array(row: impl Row) -> ListValue {
    ListValue::new(row.iter().map(|d| d.to_owned_datum()).collect())
}
```

To implement the functions in this PR, if the input types are also `impl Row`, we will have to cast the type to the inner type manually, which is verbose and dirty. Is is possible to do things like the following with the proc macro? 
```
fn generic_greatest<T>(input: impl IntoIterator<Item=Option<T>>) -> Option<T> {}
```

If not I think the current implementation is better.

Reply:
> we will have to cast the type to the inner type manually, which is verbose and dirty

You meant `T::try_from`? Anyways it would be a backward compatible change from a binary tree to a flat one later. cc @wangrunji0408

Reply:
>  Is is possible to do things like the following with the proc macro?
> ```
> fn generic_greatest<T>(input: impl IntoIterator<Item=Option<T>>) -> Option<T> {}
> ```

Yes, it's possible. But it requires some effort to support this in the macro. If this does help, I may add this support in the near future.

> we will have to cast the type to the inner type manually, which is verbose and dirty

Can we compare the `Datum` directly without downcast? I think it should be correct if all inputs have been casted to the common type.

Reply:
Btw, it seems that implicit cast to the common type is needed in the frontend. Otherwise, the backend function doesn't know how to compare values between string and int.

Reply:
> Can we compare the Datum directly without downcast?

No because there are `null first` and `null last`. For the 2 functions here we actually pick neither but ignore nulls.

https://github.com/risingwavelabs/risingwave/blob/d2ec83c88b76e2a0d9b2c9ec59b05dd43edca780/src/common/src/types/mod.rs#L547-L551

Reply:
I have implemented the variadic version with `f(row: impl Row)` and `T::try_from`. The implementation looks much better now. Thanks for the suggestions!

Reply:
Just fyi I found a case where `align_type` ([pg](https://www.postgresql.org/docs/16/typeconv-union-case.html)) is required and a binary expr tree would fail:
```
dev=> select greatest('12', '2');
 greatest 
----------
 2
(1 row)

dev=> select greatest('12', '2', 0);
 greatest 
----------
       12
(1 row)
```
The former fallback to `varchar`/`text` but the latter infers all 3 to be `int`. It is not equivalent to `greatest(greatest('12', '2'), 0)`.